Add agent skills for orchestrating cherry-picks to stable and beta channels (#187860)
flutter/flutter: .agents/skills/flutter-cherry-pick/SKILL.md

---
name: flutter-cherry-pick

> description: Cherry-pick a merged PR from master to the stable or beta channel.

Reviewer: lets add some bounds to this description. 

```suggestion
description: How to land a formal cherry-pick of a merged PR for the flutter/flutter repo stable or beta channel. Only use for flutter/flutter landed pull requests. Only use when the cherry pick request is into "stable", "beta" or a branch that has the format with flutter-<major>.<minor>-candidate.0. 
```

---

# Flutter Cherry-Pick Workflow

This skill automates the process of requesting a cherry-pick of a merged PR from the `master` branch to either the `stable` or `beta` channel.

> 

Reviewer: I know agents love to add this but all this information should be in the description. Lets remove it.

## Quick Start

Trigger this skill by providing a merged PR number or URL and optionally the target channel:
* "Cherry pick PR 187045 to stable"
* "Request CP to beta for https://github.com/flutter/flutter/pull/186952"

> * "CP PR 187045" (the agent will ask you to clarify the channel)

Reviewer: Love this section this is where we will start when we add evals.

> ## Workflows

Reviewer: This should probably start by verifying the environment is clean. On master,  branches up to date, no pending changes etc.

3. Verify you are in the root of the `flutter` repository.

### Phase 2: Run Cherry-Pick Orchestrator
1. Run the helper script to start the process:
   ```bash
   <DART_EXECUTABLE> .agents/skills/flutter-cherry-pick/scripts/flutter_cp.dart --pr <ORIGINAL_PR> --channel <CHANNEL> --action start
   ```
2. **Evaluate Script Output:**
   * **Exit Code 0 (Success):**
     - The output will end with `SUCCESS:AUTOMATED:<CP_PR>` or `SUCCESS:MANUAL:<CP_PR>`.
     - Extract the `<CP_PR>` number.
     - Proceed to **Phase 4: Fill Template**.
   * **Exit Code 2 (Conflicts):**
     - The cherry-pick encountered conflicts and the script checked out a local branch `cherry-pick-<ORIGINAL_PR>-to-<CHANNEL>`.
     - Proceed to **Phase 3: Conflict Resolution**.
   * **Exit Code 1 (Error):**
     - An unexpected error occurred. Report the failure to the user.

### Phase 3: Conflict Resolution
1. **Analyze Conflicts:**
   - Identify the conflicted files from the script output (or run `git status --porcelain` to see unmerged files).

> 2. **Attempt Auto-Resolution:**

Reviewer: Somewhere in here, maybe before attempting to resolve merge conflicts, lets have the agent articulate what needs to be resolved or what was resolved and have that the communicated in the pull request. It is good practice to know that there we conflicts when reviewing even if they were trivial.

   - For each conflicted file, view the contents (using `view_file`) and analyze the conflict markers (`<<<<<<<`, `=======`, `>>>>>>>`).
   - If the conflict is simple (e.g., import ordering, minor formatting, non-overlapping changes), resolve it using `replace_file_content` or `multi_replace_file_content`.

> 3. **User Fallback:**

Reviewer: Do you need to tell it to use these tools?

>    - If you cannot resolve the conflicts automatically:

Reviewer: ```suggestion
   - If you cannot resolve the conflicts automatically or the conflicts reflect a change in logic that impacts flutter users:
```

     - Present the list of conflicted files to the user in the chat.
     - Ask the user to resolve them in their editor.
     - Stop calling tools and wait for the user to confirm resolution.
4. **Continue the Process:**
   - Once all conflicts are resolved (either by you or the user), run `git add <resolved-files>` to stage them.
   - Resume the orchestrator script:
     ```bash
     <DART_EXECUTABLE> .agents/skills/flutter-cherry-pick/scripts/flutter_cp.dart --pr <ORIGINAL_PR> --channel <CHANNEL> --action continue
     ```
   - The script will run `git cherry-pick --continue`, push, create the PR, and print `SUCCESS:MANUAL:<CP_PR>`.
   - Extract the `<CP_PR>` number and proceed to **Phase 4: Fill Template**.

### Phase 4: Fill Template and Apply
1. **Dynamically Retrieve Template:**
   - Read the local cherry-pick template:

>      Use `view_file` on the local cherry-pick template (located at `../../../.github/PR_TEMPLATE/PULL_REQUEST_CP_TEMPLATE.md` relative to this skill)

Reviewer: make this link a markdown link so that the dart_skills_lint will validate the file exists (and fail if someone later moves or renames the file) 
```suggestion
     Use `view_file` on the local cherry-pick template (located at `[PULL_REQUEST_CP_TEMPLATE](../../../.github/PR_TEMPLATE/PULL_REQUEST_CP_TEMPLATE.md)` relative to this skill)
```

>    - If the file does not exist locally, fallback to fetching it via `read_url_content` from:

Reviewer: If the file does not exist locally why would it exist upstream?

     `https://raw.githubusercontent.com/flutter/flutter/main/.github/PR_TEMPLATE/PULL_REQUEST_CP_TEMPLATE.md`

> 2. **Gather Context:**

Reviewer: I think this section would benefit from some examples. Non blocking.
